File: resiliences-service/Services/InvestmentService.cs
using resiliences_service.Clients;
using resiliences_service.Enums;
using resiliences_service.interfaces;
using resiliences_service.Models;
using resiliences_service.Payloads;

namespace resiliences_service.Services
{
    public class InvestmentService : IInvestmentService
    {
        // ── Return rates per duration (annualized → prorated) ──────────────────
        // Weekly  → 8%  per year  → ~0.154% per week
        // Monthly → 12% per year  → 1% per month
        // Quarterly → 18% per year → 4.5% per quarter
        // Yearly  → 24% per year
        private static readonly Dictionary<InvestmentDuration, (decimal AnnualRate, int Days)> _plans = new()
        {
            { InvestmentDuration.WEEKLY,    (8m,   7)   },
            { InvestmentDuration.MONTHLY,   (12m,  30)  },
            { InvestmentDuration.QUARTERLY, (18m,  90)  },
            { InvestmentDuration.YEARLY,    (24m,  365) },
        };

        private readonly IInvestmentRepository       _repo;
        private readonly WalletServiceClient         _walletClient;
        private readonly IHistoryService             _historyService;
        private readonly ILogger<InvestmentService>  _logger;

        public InvestmentService(
            IInvestmentRepository      repo,
            WalletServiceClient        walletClient,
            IHistoryService            historyService,
            ILogger<InvestmentService> logger)
        {
            _repo           = repo;
            _walletClient   = walletClient;
            _historyService = historyService;
            _logger         = logger;
        }

        // ── helpers ────────────────────────────────────────────────────────────

        private static (decimal rate, int days) GetPlan(InvestmentDuration duration) =>
            _plans.TryGetValue(duration, out var plan)
                ? plan
                : throw new ArgumentException($"Unknown investment duration: {duration}");

        private static decimal CalculateProfit(decimal principal, decimal annualRate, int days)
        {
            // simple interest: P × r × t   (r in fraction, t in years)
            var t = days / 365m;
            return Math.Round(principal * (annualRate / 100m) * t, 4);
        }

        // ── IInvestmentService ─────────────────────────────────────────────────

        public Task<object> CalculateReturnsAsync(decimal principal, InvestmentDuration duration, string currencyCode)
        {
            var (rate, days) = GetPlan(duration);
            var profit      = CalculateProfit(principal, rate, days);
            var total       = principal + profit;

            object result = new
            {
                principal,
                duration      = duration.ToString(),
                durationDays  = days,
                annualRate    = rate,
                expectedProfit = profit,
                totalPayout   = total,
                currencyCode,
                maturityDate  = DateTime.UtcNow.AddDays(days).ToString("yyyy-MM-dd")
            };

            return Task.FromResult(result);
        }

        public async Task<Investment> CreateInvestmentAsync(CreateInvestmentRequest req)
        {
            var (rate, days) = GetPlan(req.Duration);
            var profit  = CalculateProfit(req.Principal, rate, days);
            var total   = req.Principal + profit;
            var refId   = $"INV-{req.UserId}-{DateTimeOffset.UtcNow.ToUnixTimeSeconds()}";

            // 1. Debit the user's main wallet
            bool debited = await _walletClient.DebitForInvestmentAsync(
                req.UserId, req.WalletId, req.CurrencyCode, req.Principal, refId);

            if (!debited)
                throw new InvalidOperationException("Insufficient wallet balance or wallet service unavailable.");

            // 2. Persist the investment record
            var investment = new Investment
            {
                UserId        = req.UserId,
                WalletId      = req.WalletId,
                CurrencyCode  = req.CurrencyCode,
                Principal     = req.Principal,
                ReturnRate    = rate,
                ExpectedProfit = profit,
                TotalPayout   = total,
                Duration      = req.Duration,
                DurationDays  = days,
                StartDate     = DateTime.UtcNow,
                MaturityDate  = DateTime.UtcNow.AddDays(days),
                Status        = InvestmentStatus.ACTIVE,
                ReferenceId   = refId
            };

            var created = await _repo.CreateAsync(investment);

            // 3. Write a transaction history record
            try
            {
                await _historyService.CreateDepositAsync(new History
                {
                    UserId           = (ulong)req.UserId,
                    WalletId         = (ulong)req.WalletId,
                    CurrencyType     = req.CurrencyCode,
                    Type             = TransactionType.DEBITED,
                    GrossAmount      = req.Principal,
                    NetAmount        = req.Principal,
                    Description      = $"Investment locked — {req.Duration} plan at {rate}% p.a. (ref: {refId})",
                    Status           = "SUCCESS",
                    ReferenceId      = refId,
                    TransactionId    = $"TXN_{Guid.NewGuid():N}"[..20].ToUpper(),
                    SessionId        = $"SESS_{Guid.NewGuid():N}"[..20].ToUpper(),
                    Timestamp        = DateTime.UtcNow
                });
            }
            catch (Exception ex)
            {
                _logger.LogWarning(ex, "[InvestmentService] History record failed for investment {Id}", created.Id);
            }

            return created;
        }

        public async Task<Investment?> GetByIdAsync(long id, long userId)
        {
            var inv = await _repo.FindByIdAsync(id);
            return (inv == null || inv.UserId != userId) ? null : inv;
        }

        public async Task<List<Investment>> GetByUserIdAsync(long userId) =>
            await _repo.FindByUserIdAsync(userId);

        /// <summary>
        /// Called periodically by the InvestmentMaturityWorker.
        /// Finds all ACTIVE investments whose maturity date has passed and credits
        /// principal + profit back to the user's wallet.
        /// </summary>
        public async Task ProcessMaturedInvestmentsAsync()
        {
            var matured = await _repo.FindMaturedUnpaidAsync();
            _logger.LogInformation("[InvestmentService] Processing {Count} matured investments", matured.Count);

            foreach (var inv in matured)
            {
                try
                {
                    var refId = $"INV-PAYOUT-{inv.UserId}-{DateTimeOffset.UtcNow.ToUnixTimeSeconds()}";

                    bool credited = await _walletClient.CreditInvestmentPayoutAsync(
                        inv.UserId, inv.WalletId, inv.CurrencyCode, inv.TotalPayout, refId);

                    if (!credited)
                    {
                        _logger.LogError("[InvestmentService] Payout failed for investment {Id}", inv.Id);
                        inv.Status = InvestmentStatus.FAILED;
                        await _repo.UpdateAsync(inv);
                        continue;
                    }

                    inv.Status    = InvestmentStatus.PAID_OUT;
                    inv.PaidOutAt = DateTime.UtcNow;
                    await _repo.UpdateAsync(inv);

                    // Write history
                    try
                    {
                        await _historyService.CreateCreditAsync(new History
                        {
                            UserId        = (ulong)inv.UserId,
                            WalletId      = (ulong)inv.WalletId,
                            CurrencyType  = inv.CurrencyCode,
                            Type          = TransactionType.CREDITED,
                            GrossAmount   = inv.TotalPayout,
                            NetAmount     = inv.TotalPayout,
                            Description   = $"Investment matured — principal {inv.Principal:F2} + profit {inv.ExpectedProfit:F2} (ref: {refId})",
                            Status        = "SUCCESS",
                            ReferenceId   = refId,
                            TransactionId = $"TXN_{Guid.NewGuid():N}"[..20].ToUpper(),
                            SessionId     = $"SESS_{Guid.NewGuid():N}"[..20].ToUpper(),
                            Timestamp     = DateTime.UtcNow
                        });
                    }
                    catch (Exception ex)
                    {
                        _logger.LogWarning(ex, "[InvestmentService] Payout history record failed for investment {Id}", inv.Id);
                    }

                    _logger.LogInformation("[InvestmentService] Payout processed for investment {Id}, user {UserId}, amount {Amount}",
                        inv.Id, inv.UserId, inv.TotalPayout);
                }
                catch (Exception ex)
                {
                    _logger.LogError(ex, "[InvestmentService] Unexpected error processing investment {Id}", inv.Id);
                }
            }
        }
    }
}

File: investment/src/main/java/com/epay/investment/service/InvestmentService.java
package com.epay.investment.service;

import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;
import com.epay.common.exception.BadRequestException;
import com.epay.common.exception.ErrorCode;
import com.epay.common.exception.ResourceNotFoundException;
import com.epay.common.exception.WalletException;
import com.epay.common.interfaces.InvestmentServices;
import com.epay.domain.history.entity.Transaction;
import com.epay.domain.history.enums.TransactionStatus;
import com.epay.domain.investment.dto.InvestmentReturnResponse;
import com.epay.domain.investment.entity.Investment;
import com.epay.domain.investment.enums.InvestmentDuration;
import com.epay.domain.investment.enums.InvestmentStatus;
import com.epay.domain.investment.enums.TransactionType;
import com.epay.domain.investment.input.CreateInvestmentRequest;
import com.epay.domain.investment.repository.InvestmentRepository;
import com.epay.domain.wallet.entity.CurrencyBalance;
import com.epay.domain.wallet.entity.Wallet;
import com.epay.history.repository.TransactionRepository;
import com.epay.wallet.repository.WalletRepository;
import java.math.BigDecimal;
import java.math.RoundingMode;
import java.time.Instant;
import java.time.LocalDate;
import java.time.LocalDateTime;
import java.util.EnumMap;
import java.util.List;
import java.util.Map;
import java.util.UUID;

@Slf4j
@Service
@RequiredArgsConstructor
public class InvestmentService implements InvestmentServices {

        private static final Map<InvestmentDuration, InvestmentPlan> PLANS = new EnumMap<>(InvestmentDuration.class);

        static {
                PLANS.put(InvestmentDuration.WEEKLY,
                        new InvestmentPlan(BigDecimal.valueOf(8), 7));

                PLANS.put(InvestmentDuration.MONTHLY,
                        new InvestmentPlan(BigDecimal.valueOf(12), 30));

                PLANS.put(InvestmentDuration.QUARTERLY,
                        new InvestmentPlan(BigDecimal.valueOf(18), 90));

                PLANS.put(InvestmentDuration.YEARLY,
                        new InvestmentPlan(BigDecimal.valueOf(24), 365));
        }

        private final InvestmentRepository investmentRepository;
        private final WalletRepository walletRepository;
        private final TransactionRepository historyService;

        private InvestmentPlan getPlan(InvestmentDuration duration) {
                InvestmentPlan plan = PLANS.get(duration);

                if (plan == null) {
                throw new IllegalArgumentException(
                        "Unknown investment duration: " + duration);
                }

                return plan;
        }

        private BigDecimal calculateProfit(
                BigDecimal principal,
                BigDecimal annualRate,
                int days
        ) {

                BigDecimal years = BigDecimal.valueOf(days)
                        .divide(BigDecimal.valueOf(365), 8, RoundingMode.HALF_UP);

                return principal
                        .multiply(annualRate.divide(BigDecimal.valueOf(100), 8, RoundingMode.HALF_UP))
                        .multiply(years)
                        .setScale(4, RoundingMode.HALF_UP);
        }


        @Override
        public InvestmentReturnResponse calculateReturns(
                BigDecimal principal,
                InvestmentDuration duration,
                String currencyCode
        ) {

                InvestmentPlan plan = getPlan(duration);

                BigDecimal profit = calculateProfit(
                        principal,
                        plan.annualRate(),
                        plan.days());

                BigDecimal total = principal.add(profit);

                return InvestmentReturnResponse.builder()
                        .principal(principal)
                        .duration(duration.name())
                        .durationDays(plan.days())
                        .annualRate(plan.annualRate())
                        .expectedProfit(profit)
                        .totalPayout(total)
                        .currencyCode(currencyCode)
                        .maturityDate(LocalDate.now().plusDays(plan.days()))
                        .build();
        }

        @Override
        @Transactional
        public Investment createInvestment(CreateInvestmentRequest request) {
                if (request.getUserId() == null){
                        throw new BadRequestException("User ID is required", ErrorCode.INVALID_INPUT);
                }

                if (request.getWalletId() == null){
                        throw new BadRequestException("Wallet ID is required", ErrorCode.INVALID_INPUT);
                }

                if (request.getCurrencyCode() == null || request.getCurrencyCode().isBlank()) {
                        throw new BadRequestException("Currency code is required", ErrorCode.INVALID_INPUT);
                }

                if (request.getPrincipal() == null || request.getPrincipal().compareTo(BigDecimal.ZERO) <= 0) {
                        throw new BadRequestException("Investment amount must be greater than zero.", ErrorCode.INVALID_INPUT);
                }

                if (request.getDuration() == null) {
                throw new BadRequestException("Investment duration is required.", ErrorCode.INVALID_INPUT);  
                }

                Wallet wallet = walletRepository.findById(request.getWalletId())
                .orElseThrow(() ->
                        new IllegalArgumentException("Wallet not found."));

                if (!wallet.getUserId().equals(request.getUserId())) {
                        throw new BadRequestException("Wallet does not belong to this user.", ErrorCode.WALLET_NOT_FOUND);                
                }

                CurrencyBalance balance = wallet.getBalance(request.getCurrencyCode())
                .orElseThrow(() -> new ResourceNotFoundException("Wallet does not support currency " + request.getCurrencyCode()));

                BigDecimal previousBalance = balance.getBalance();

                if (previousBalance.compareTo(request.getPrincipal()) < 0)
                throw new WalletException(
                        String.format("Insufficient %s balance. Available: %.2f",
                                request.getCurrencyCode(), balance.getBalance()), ErrorCode.INSUFFICIENT_BALANCE);

                
                InvestmentPlan plan = getPlan(request.getDuration());

                BigDecimal profit = calculateProfit(
                        request.getPrincipal(),
                        plan.annualRate(),
                        plan.days());

                BigDecimal total = request.getPrincipal().add(profit);

                String referenceId =
                        "INV-" + request.getUserId() + "-" + Instant.now().getEpochSecond();

                BigDecimal newBalance = previousBalance.subtract(request.getPrincipal());

                balance.setBalance(newBalance);

                walletRepository.save(wallet);
                
                Investment investment = Investment.builder()
                        .userId(request.getUserId())
                        .walletId(wallet.getId())
                        .currencyCode(request.getCurrencyCode())
                        .principal(request.getPrincipal())
                        .returnRate(plan.annualRate())
                        .expectedProfit(profit)
                        .totalPayout(total)
                        .duration(request.getDuration())
                        .durationDays(plan.days())
                        .startDate(LocalDateTime.now())
                        .maturityDate(LocalDateTime.now().plusDays(plan.days()))
                        .status(InvestmentStatus.ACTIVE)
                        .referenceId(referenceId)
                        .build();

                investment = investmentRepository.save(investment);
        try {
                Transaction transaction = Transaction.builder()
                        .transactionId(generateId("TXN"))
                        .reference(referenceId)
                        .userId(request.getUserId())
                        .walletId(wallet.getId())
                        .transactionType(TransactionType.DEBITED.name())
                        .debitCredit("DEBIT")
                        .currency(request.getCurrencyCode())
                        .grossAmount(request.getPrincipal())
                        .netAmount(request.getPrincipal())
                        .previousBalance(previousBalance)
                        .availableBalance(newBalance)
                        .runningBalance(newBalance)
                        .description(String.format(
                                "Investment locked (%s plan @ %s%% p.a.)",
                                request.getDuration(),
                                plan.annualRate()))
                        .build();
                        transaction.advanceStatus(TransactionStatus.INITIATED, "SYSTEM", "Transaction created");
                        transaction.advanceStatus(TransactionStatus.SETTLED, "SYSTEM", "Investment created successfully.");

                        historyService.save(transaction);
                } catch (Exception ex) {
                        log.warn("Failed to save investment transaction history for investment {}", investment.getId(), ex);
                }
                return investment;
        }

        @Override
        public Investment getById(Long id, Long userId) {
                return investmentRepository.findById(id)
                        .filter(i -> i.getUserId().equals(userId))
                        .orElse(null);

        }

        @Override
        public List<Investment> getByUserId(Long userId) {
                return investmentRepository.findByUserId(userId);
        }

        @Override
        @Transactional
        public void processMaturedInvestments() {

                List<Investment> matured = investmentRepository.findMaturedUnpaid(LocalDateTime.now());

                log.info("Processing {} matured investments", matured.size());

                for (Investment investment : matured) {

                        try {
                        String reference ="INV-PAYOUT-" + investment.getUserId() +"-" + Instant.now().getEpochSecond();

                        Wallet wallet = walletRepository.findById(investment.getWalletId())
                                .orElseThrow(() ->
                                        new IllegalStateException(
                                                "Wallet not found for investment " + investment.getId()));

                        CurrencyBalance balance = wallet.getBalance(investment.getCurrencyCode())
                                .orElseThrow(() ->
                                        new IllegalStateException(
                                                "Currency " + investment.getCurrencyCode() + " not found in wallet."));

                        BigDecimal previousBalance = balance.getBalance();

                        BigDecimal newBalance =
                                previousBalance.add(investment.getTotalPayout());

                        balance.setBalance(newBalance);

                        walletRepository.save(wallet);

                        investment.setStatus(InvestmentStatus.PAID_OUT);
                        investment.setPaidOutAt(LocalDateTime.now());

                        investmentRepository.save(investment);

                        try {
                                Transaction transaction = Transaction.builder()
                                        .transactionId(generateId("TXN"))
                                        .reference(reference)
                                        .userId(investment.getUserId())
                                        .walletId(investment.getWalletId())
                                        .transactionType(TransactionType.CREDITED.name())
                                        .debitCredit("CREDIT")
                                        .currency(investment.getCurrencyCode())
                                        .grossAmount(investment.getTotalPayout())
                                        .netAmount(investment.getTotalPayout())
                                        .previousBalance(previousBalance)
                                        .availableBalance(newBalance)
                                        .runningBalance(newBalance)
                                        .description(String.format(
                                                "Investment matured — principal %s + profit %s",
                                                investment.getPrincipal(),
                                                investment.getExpectedProfit()))
                                        .build();

                                transaction.advanceStatus(TransactionStatus.INITIATED,"SYSTEM","Transaction created");
                                transaction.advanceStatus(TransactionStatus.SETTLED,"SYSTEM","Investment payout completed.");

                                historyService.save(transaction);

                        } catch (Exception ex) {

                                log.warn(
                                        "Failed to write payout history for investment {}",
                                        investment.getId(),
                                        ex);

                        }

                        log.info(
                                "Investment {} paid successfully. User={}, Amount={}",
                                investment.getId(),
                                investment.getUserId(),
                                investment.getTotalPayout());

                        } catch (Exception ex) {

                        investment.setStatus(InvestmentStatus.FAILED);
                        investmentRepository.save(investment);

                        log.error(
                                "Failed to process matured investment {}",
                                investment.getId(),
                                ex);

                        }

                }
        }

        private String generateId(String prefix) {

                return (prefix + "_" +
                        UUID.randomUUID()
                                .toString()
                                .replace("-", "")
                                .toUpperCase())
                        .substring(0, 20);

        }

        private record InvestmentPlan(BigDecimal annualRate,int days) {
        }
}